{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "772feefd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 7 datasets.\n"
     ]
    }
   ],
   "source": [
    "from BCBio import GFF\n",
    "from reportlab.lib.units import cm\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import os\n",
    "\n",
    "# Load datasets and feature files (gff) in pairs\n",
    "classification_file = Path(\"../data/classified\")\n",
    "feature_files = Path(\"../data/features\")\n",
    "datasets = []\n",
    "output = Path(\"../data/genomecharts\")\n",
    "output.mkdir(exist_ok=True)\n",
    "\n",
    "for class_file in classification_file.glob(\"*.tsv\"):\n",
    "    # Grab dataset name\n",
    "    set_name = class_file.name.split(\"_\")[0].lower()\n",
    "    # Special cases for guegler dataset\n",
    "    if set_name == \"guegler\" and class_file.name.split(\"_\")[1].lower() == \"t4\":\n",
    "        set_name = \"gueglert4\"\n",
    "    elif set_name == \"guegler\" and class_file.name.split(\"_\")[1].lower() == \"t7\":\n",
    "        set_name = \"gueglert7\"\n",
    "\n",
    "    # Get feature (gff) file\n",
    "    for feature_file in feature_files.glob(\"*.gff3\"):\n",
    "        name_split = feature_file.name.split(\"_\")\n",
    "        set_feature_name = name_split[0].lower()\n",
    "\n",
    "        if set_name == set_feature_name:\n",
    "            datasets.append((set_name, class_file, feature_file))\n",
    "\n",
    "print(f\"Found {len(datasets)} datasets.\")\n",
    "\n",
    "\n",
    "def parse_input_files(classification_file, gff_file):\n",
    "    \"\"\"\n",
    "    Parse a gff feature file and classification table.\n",
    "\n",
    "    Returns classification table as a dataframe and gff as a SeqIO record.\n",
    "    \"\"\"\n",
    "    in_file = gff_file\n",
    "    class_file = classification_file\n",
    "    df = pd.read_csv(class_file, sep=\"\\t\")\n",
    "    in_handle = open(in_file)\n",
    "\n",
    "    for rec in GFF.parse(in_handle):\n",
    "        record = rec\n",
    "    in_handle.close()\n",
    "\n",
    "    return df, record\n",
    "\n",
    "\n",
    "def draw_genechat(set_name, df, record, feature_basename):\n",
    "    \"\"\"\n",
    "    Draw a gene chart for a genome, classifed in the dataframe, with the features from a seqIO record\n",
    "    \"\"\"\n",
    "    gd_diagram = GenomeDiagram.Diagram(f\"{set_name}\")\n",
    "    gd_track_for_features = gd_diagram.new_track(1, name=\"Annotated Features\")\n",
    "    gd_feature_set = gd_track_for_features.new_set()\n",
    "\n",
    "    colormap = {\n",
    "        \"early\": \"#fee8c8\",\n",
    "        \"middle\": \"#fdbb84\",\n",
    "        \"late\": \"#e34a33\",\n",
    "        \"None\": \"grey\",\n",
    "    }\n",
    "\n",
    "    # Add genome start marker as a fake feature at position 0\n",
    "    start_marker = SeqFeature(FeatureLocation(0, 1), type=\"marker\")\n",
    "    gd_feature_set.add_feature(\n",
    "        start_marker,\n",
    "        color=\"blue\",\n",
    "        label_position=\"start\",\n",
    "        label_size=10,\n",
    "    )\n",
    "\n",
    "    for feature in record.features:\n",
    "        if feature.type != \"gene\":\n",
    "            continue\n",
    "\n",
    "        classification = df.loc[df[\"Geneid\"] == feature.id, \"classification\"]\n",
    "        classification = (\n",
    "            classification.values[0] if not classification.empty else \"None\"\n",
    "        )\n",
    "        color = colormap[classification]\n",
    "        gd_feature_set.add_feature(feature, color=color, label=False)\n",
    "\n",
    "    # Add a separate legend track for legend\n",
    "    legend_track = gd_diagram.new_track(2, name=\"Legend\", scale=False, height=0.5)\n",
    "    legend_set = legend_track.new_set()\n",
    "\n",
    "    # Add a dummy feature per classification as legend\n",
    "    legend_pos = 0\n",
    "    c = 800\n",
    "    for label, color in colormap.items():\n",
    "        dummy_feature = SeqFeature(\n",
    "            FeatureLocation(legend_pos, legend_pos + c), type=\"legend\"\n",
    "        )\n",
    "        legend_set.add_feature(\n",
    "            dummy_feature,\n",
    "            color=color,\n",
    "            label=True,\n",
    "            label_size=10,\n",
    "            name=label,\n",
    "        )\n",
    "        legend_pos += c  # space out legend items\n",
    "\n",
    "    gd_diagram.draw(\n",
    "        format=\"circular\",\n",
    "        circular=True,\n",
    "        pagesize=(20 * cm, 20 * cm),\n",
    "        start=0,\n",
    "        end=len(record),\n",
    "        circle_core=0.7,\n",
    "    )\n",
    "    gd_diagram.write(f\"{output}/{feature_basename}_genechart.png\", \"PNG\")\n",
    "\n",
    "\n",
    "# Draw genome chart for each phage in datasets\n",
    "for name, df_class, gff_file in datasets:\n",
    "    df, record = parse_input_files(df_class, gff_file)\n",
    "    feature_basename = gff_file.stem  # get full filename\n",
    "    draw_genechat(name, df, record, feature_basename)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vividVirions",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
